marray/__init__.py

```diff
@@ -79,20 +79,26 @@ def __setitem__(self, key, other):
             self.mask[key] = getattr(other, 'mask', False)
             return self.data.__setitem__(key, getattr(other, 'data', other))
 
+        def _data_mask_string(self, fun):
+            data_str = fun(self.data) + ", "
+            mask_str = fun(self.mask)
+            if len(data_str) + len(mask_str) <= 66:
+                join_char = ""
+                components = ["MaskedArray(", data_str, mask_str, ")"]
+            else:
+                join_char = "\n"
+                components = ["MaskedArray(",
+                              textwrap.indent(data_str, " " * 4),
+                              textwrap.indent(mask_str, " " * 4),
+                              ")"]
+            return join_char.join(components)
+
         ## Visualization ##
         def __repr__(self):
-            # temporary: fix for CuPy
-            # eventually: rewrite to avoid masked array
-            data = np.asarray(self.data)
-            mask = np.asarray(self.mask)
-            return np.ma.masked_array(data, mask).__repr__()
+            return self._data_mask_string(repr)
 
         def __str__(self):
-            # temporary: fix for CuPy
-            # eventually: rewrite to avoid masked array
-            data = np.asarray(self.data)
-            mask = np.asarray(self.mask)
-            return np.ma.masked_array(data, mask).__str__()
+            return self._data_mask_string(str)
```

Comment:
I wonder if this should always be `repr`? That would mean there's no difference between `__str__` and `__repr__`, but given that at least `numpy` is always a pseudo-`repr` anyways (so recreating objects is not possible) this might not be too much of an issue?

Reply:
> I wonder if this should always be repr?

But why _not_ follow the convention of the underlying library, whatever it is?

```python3
import numpy as np
import torch
import jax.numpy as jnp
import cupy as cp
for xp in [np, torch, jnp, cp]:
    x = xp.asarray([1, 2, 3, 4])
    print(xp.__name__)
    print(str(x))
    print(repr(x))

# numpy
# [1 2 3 4]
# array([1, 2, 3, 4])
#
# torch
# tensor([1, 2, 3, 4])
# tensor([1, 2, 3, 4])
#
# jax.numpy
# [1 2 3 4]
# Array([1, 2, 3, 4], dtype=int32)
#
# cupy
# [1 2 3 4]
# array([1, 2, 3, 4])
```
Most of them have different `str` vs `repr`, so I don't see a reason not to do what they do.

> but given that at least numpy is always a pseudo-repr anyways (so recreating objects is not possible) this might not be too much of an issue?

I don't think it's much of an issue, especially since this is just a placeholder until it's time to polish things.
But if we were to expose `MaskedArray` (and maybe accept `dtype` as a parameter), then with the `repr` output, the object could be recreated to the extent that the NumPy `array` object can be recreated.